File: code/binding/development/tools/cast/src/java/cast/architecture/ManagedComponent.java
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

/**
 * 
 */
package cast.architecture;

import java.util.HashMap;

import Ice.Current;
import cast.cdl.TaskManagementDecision;
import cast.cdl.TaskOutcome;
import cast.core.Pair;
import cast.core.QueuedDataRunnable;
import cast.interfaces.TaskManagerPrx;
import cast.interfaces._ManagedComponentOperations;

/**
 * An abstract class to represent a component in a subarchitecture that can read
 * and write to a working memory, and has its operations controlled by a task
 * manager.
 * 
 * @author nah
 */
public abstract class ManagedComponent extends WorkingMemoryReaderComponent
		implements _ManagedComponentOperations {

	private static class TaskManagementResult {
		public String m_id;
		public TaskManagementDecision m_decision;

		public TaskManagementResult(String _id, TaskManagementDecision _decision) {
			m_id = _id;
			m_decision = _decision;
		}
	}

	private TaskManagerPrx m_taskManager;
	private final HashMap<String, Pair<String, String>> m_proposedTasks;

	public ManagedComponent() {
		m_proposedTasks = new HashMap<String, Pair<String, String>>();
		m_taskCounter = 0;
		m_taskNotificationLock = new Object();
		m_notifications = 0;
		m_responseRunnable = null;
	}

	public void setTaskManager(TaskManagerPrx _tm, Current __current) {
		m_taskManager = _tm;
	}

	/**
	 * Inner class used to forward the responses of the task manager on to the
	 * subclass.
	 * 
	 * @author nah
	 */
	private static class GMResponseHandler extends
			QueuedDataRunnable<TaskManagementResult> {

		private ManagedComponent m_gdp;

		/**
		 * @param _gdp
		 * @param _goalDecision
		 * @param _goalID
		 */
		public GMResponseHandler(ManagedComponent _gdp) {
			super();
			m_gdp = _gdp;
		}

		@Override
		protected void nextInQueue(TaskManagementResult _data) {
			m_gdp.lockComponent();
			m_gdp.handleTaskManagerResponse(_data.m_id, _data.m_decision);
			m_gdp.unlockComponent();
		}

	}

	private final Object m_taskNotificationLock;

	/**
	 * Count of notifications that are outstanding in the "wait" cycle.
	 */
	int m_notifications;

	/**
	 * Counter for generating task ids.
	 */
	protected int m_taskCounter;

	private GMResponseHandler m_responseRunnable;

	/**
	 * Receive a response from the task manager and call the correct
	 * result-specific method.
	 * 
	 * @param _taskID
	 *            The id of the task.
	 * @param _decision
	 *            The task manager decision.
	 */
	private void handleTaskManagerResponse(String _taskID,
			TaskManagementDecision _decision) {

		// inform sub-class of result
		if (_decision.value() == TaskManagementDecision.TaskAdopted.value()) {
			taskAdopted(_taskID);
			synchronized (m_taskNotificationLock) {
				m_notifications++;
				m_taskNotificationLock.notifyAll();
			}

		} else if (_decision.value() == TaskManagementDecision.TaskRejected.value()
				) {
			taskRejected(_taskID);
			synchronized (m_taskNotificationLock) {
				m_notifications++;
				m_taskNotificationLock.notifyAll();
			}

		} else if (_decision.value() == TaskManagementDecision.TaskWaiting.value()) {
			debug("TaskWaiting received for: " + _taskID);
		} 

	}

	/**
	 * Generate a new unique task id.
	 * 
	 * @return A unique task id.
	 */
	protected String newTaskID() {
		StringBuffer sb = new StringBuffer("task:");
		sb.append(getComponentID());
		sb.append(m_taskCounter++);
		return sb.toString();
	}

	/**
	 * Propose the processing of a task.
	 * 
	 * @param _taskID
	 *            The id used to refer to this task.
	 * @param _taskName
	 *            The name of the task.
	 */
	protected void proposeInformationProcessingTask(String _taskID,
			String _taskName) {

		assert (_taskID != null);
		assert (_taskName != null);
		assert (m_taskManager != null);

		// store it as proposed
		m_proposedTasks.put(_taskID, new Pair<String, String>(_taskID,
				_taskName));

		m_taskManager.proposeTask(getComponentID(), _taskID, _taskName);
	}

	// /**
	// * Register a list of tasks that this component can perform.
	// *
	// * @param _taskDesc
	// * An array of task descriptions.
	// * @throws SubarchitectureProcessException
	// * if connection to goal manager is not set or on comms error.
	// */
	// protected void registerTaskDescriptions(TaskDescription[] _taskDesc)
	// throws SubarchitectureProcessException {
	//
	// if (m_taskDescriptionRegistration == null) {
	// throw new SubarchitectureProcessException(
	// "Push connection not to sa goal manager not set.");
	// }
	//
	// // push registration to s-a goal manager
	// m_taskDescriptionRegistration.push(getProcessIdentifier().toString(),
	// _taskDesc);
	// }

	/**
	 * Retract the proposal of processing a task.
	 * 
	 * @param _taskID
	 *            The id of the task.
	 * @throws SubarchitectureProcessException
	 *             if connection to goal manager is not set or on comms error.
	 */
	protected void retractInformationProcessingTask(String _taskID) {
		assert (m_taskManager != null);
		m_proposedTasks.remove(_taskID);
		m_taskManager.retractTask(getComponentID(),_taskID);
	}

	/**
	 * Receive a signal that an information processing task has been accepted.
	 * 
	 * @param _taskID
	 */
	protected void taskAdopted(String _taskID) {

	}

	/**
	 * Signal the outcome of goal processing to the sub-architecture goal
	 * manager.
	 * 
	 * @param _taskID
	 *            The task id.
	 * @param _outcome
	 *            The task outcome.
	 * @throws SubarchitectureProcessException
	 *             if connection to goal manager is not set.`
	 */
	protected void taskComplete(String _taskID, TaskOutcome _outcome) {

		assert (_taskID != null);
		assert (_outcome != null);
		assert (m_taskManager != null);

		// remove the goal from the hashtable
		m_proposedTasks.remove(_taskID);

		m_taskManager.taskComplete(getComponentID(),_taskID, _outcome);
	}

	/**
	 * Receive a signal that an information processing goal has been rejected.
	 * 
	 * @param _taskID
	 */
	protected void taskRejected(String _taskID) {

	}
	

	
	//
	/**
	 * Put the process to sleep until a new task notification is received. This
	 * method will sleep until either the taskAdopted or taskRejected functions
	 * next <b>exit</b>, unless these methods have exited more times than this
	 * method has been called. In this case the method does not block, as the
	 * assumption is that the process has to process these outstanding
	 * notifications first. This latter behaviour was added to fixed bug #31.
	 * 
	 * @remark Interface change: Used to take a single argument. This was
	 *         removed to simplify behaviour.
	 */
	protected void waitForNotifications() {

		synchronized (m_taskNotificationLock) {
			if (m_notifications == 0) {
				// println("blocking for new");
				try {
					m_taskNotificationLock.wait();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
				// the count will have been incremented in the same action as
				// the broadcast
				m_notifications--;
				assert (m_notifications >= 0);

			}
			// else decrement the cout
			else {
				m_notifications--;
				assert (m_notifications >= 0);
			}
		}

	}

	public void taskDecision(String _id, TaskManagementDecision _decision, Current __current) {

		debug("received goal management decision: " + _id);

		// if this goal was proposed by this component
		// hashtable is synchronised, so no need to protect it
		if (m_proposedTasks.containsKey(_id)) {

			if (m_responseRunnable == null) {
				m_responseRunnable = new GMResponseHandler(this);
				m_responseRunnable.start();
				new Thread(m_responseRunnable).start();
			}

			m_responseRunnable.queue(new TaskManagementResult(_id, _decision));
		}
	}

	@Override
	public void stopInternal() {

		if (m_responseRunnable != null) {
			m_responseRunnable.stop();
		}

		synchronized (m_taskNotificationLock) {
			m_taskNotificationLock.notifyAll();
		}
	}

}

File: binding/development/tools/cast/src/c++/cast/architecture/ManagedComponent.cpp
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

#include "ManagedComponent.hpp"
#include <sstream>
#include <exception>

using namespace std;

namespace cast {

  ManagedComponent::ManagedComponent() :
    m_notifications(0),
    //    m_pProposedTasks(new IPTaskMap()),
    m_taskCounter(0)// ,
//     m_responseRunnable(NULL)
  {
    
//     m_pTaskNotificationCondition 
//       = new omni_condition(&m_taskNotificationMutex);

  }

  ManagedComponent::~ManagedComponent() {


//     QueuedDataThread<cdl::TaskManagementResult>::cleanup(m_responseRunnable);

//     m_taskNotificationMutex.lock();
//     //dummy value to prevent assert on exit
//     m_notifications = 100;
//     m_pTaskNotificationCondition->broadcast();
//     m_taskNotificationMutex.unlock();
//     delete m_pTaskNotificationCondition;

//     //TODO might need to empty it too...
//     delete m_pProposedTasks;
  }

  void ManagedComponent::stopInternal() {
    //FrameworkComponent::

//     //stop threads
//     if(m_responseRunnable != NULL) {
//       m_responseRunnable->stop();
//     }


//     //release sleeping threads
//     m_taskNotificationMutex.lock();
//     m_pTaskNotificationCondition->broadcast();
//     m_taskNotificationMutex.unlock();
    WorkingMemoryReaderComponent::stopInternal();
    
  }

  void ManagedComponent::waitForNotifications() {

    //    m_taskNotificationMutex.lock();    
//     //if there are no outstanding notifications, then block
//     if(m_notifications == 0) {
//       //println("blocking for new");
//       m_pTaskNotificationCondition->wait();
//       //the count will have been incremented in the same action as the broadcast
//       m_notifications--;
//       assert(m_notifications >= 0);

//     }
//     //else decrement the coutn
//     else {
//       m_notifications--;
//       assert(m_notifications >= 0);
//     }
//     m_taskNotificationMutex.unlock();	            
    

  }




//   void ManagedComponent::receivePushData(FrameworkLocalData<cdl::TaskManagementResult> *_pData) {


//     //protect access to data structures
//     m_gdpMutex.lock();


//     try {


//       if(m_status == STATUS_RUN) {

// 	//check that the decision is for this component
// 	string taskID(_pData->data()->m_id);
// 	IPTaskMap::iterator i = m_pProposedTasks->find(taskID);
	
// 	//printfln("received decision!!, %s", taskID.c_str());

// 	//if the task is one we proposed
// 	if(i != m_pProposedTasks->end()) {
	  
// 	  if(m_responseRunnable == NULL) { 	    
// 	    m_responseRunnable = new GMResponseHandler(this);
// 	    m_responseRunnable->start(;
// 	  }
	
// 	  //copy ;)
// 	  //	  println("queued!!");
// 	  m_responseRunnable->queue(*_pData->data());
    
// 	} else {
// 	  //not for us! ... 
// 	  //println("shit!!");
// 	}
//       }


//     }
//     catch (exception& e)
//       {
// 	println("exception details follow...");
// 	cerr << e.what() << endl;
//       }
//     catch (...) {
//       println("exception details follow...");
//       cerr << "default exception" << endl;
//     }

//     m_gdpMutex.unlock();
  
//     // data needs to be deleted whatever happens
//     delete _pData;


//   }


  string ManagedComponent::newTaskID() {
    std::ostringstream o;
    o << "IPG:";
    o << getComponentID();
    o << ":";
    o << m_taskCounter;

    m_taskCounter++;

    return o.str();    
  }


//   void ManagedComponent::proposeInformationComponentingTask(const string &_taskID, const string &_taskName) {

//     if(m_pComponentingResultToGoalManager) {

//       logTaskProposed(_taskID,_taskName);

//       //create the goal structure
//       cdl::InformationComponentingTask *pIPT 
// 	= new cdl::InformationComponentingTask();
//       pIPT->m_id = CORBA::string_dup(_taskID.c_str());
//       pIPT->m_taskName =  CORBA::string_dup(_taskName.c_str());

//       //store it as proposed
//       (*m_pProposedTasks)[_taskID] = pIPT;

//       //COPY (not point) data into local data object... as it will
//       //get deleted by the sending component
//       FrameworkLocalData<cdl::InformationComponentingTask> *pFLD 
// 	= new FrameworkLocalData<cdl::InformationComponentingTask>(getComponentID(),*pIPT);
    
//       // push task to s-a goal manager
//       m_pProposalInputToTaskManager->push(pFLD);
      
//     }
//     else {
//       throw(SubarchitectureComponentException(__HERE__,"Push connection not to sa goal manager for proposal not set."));
//     }

//   }

//   void ManagedComponent::retractInformationComponentingTask(const string &_taskID) {

//     if(m_pRetractionInputToTaskManager) {

//       FrameworkLocalData<std::string> *pFLD 
// 	= new FrameworkLocalData<std::string>(getComponentID(),new string(_taskID));
    
//       // push task to s-a goal manager
//       m_pRetractionInputToTaskManager->push(pFLD);
      
//     }
//     else {
//       throw(SubarchitectureComponentException(__HERE__,"Push connection not to sa goal manager for retraction not set."));
//     }

//   }

//   void ManagedComponent::taskComplete(const string &_goalID, 
// 				    const cdl::TaskOutcome &_outcome) {

//     if (m_pComponentingResultToGoalManager) {

//       logTaskComplete(_goalID);

//       IPTaskMap::iterator i = m_pProposedTasks->find(_goalID);
//       if(i != m_pProposedTasks->end()) {
// 	m_pProposedTasks->erase(i);
//       }


//       cdl::TaskResult *pTR = new cdl::TaskResult();
//       pTR->m_id = CORBA::string_dup(_goalID.c_str());
//       pTR->m_outcome = _outcome;

//       //use our memory for the data, but the FLD class will delete when
//       //sending is complete
//       FrameworkLocalData<cdl::TaskResult> *pFLD 
// 	= new FrameworkLocalData<cdl::TaskResult>(getComponentID(),
// 						  pTR);
//       m_pComponentingResultToGoalManager->push(pFLD);
//     }
//     else {
//       throw(SubarchitectureComponentException(__HERE__,"Goal componenting result connection to goal manager has not been set up."));
//     }
//   }

//   void ManagedComponent::registerTaskDescriptions(cdl::TaskDescriptionList * _pTaskList) {

//     if (m_pTaskDescriptionRegistration) {
    
//       FrameworkLocalData<cdl::TaskDescriptionList> *pFLD 
// 	= new FrameworkLocalData<cdl::TaskDescriptionList>(getComponentID(),
// 							   _pTaskList);
//       m_pTaskDescriptionRegistration->push(pFLD);
//     }
//     else {
//       throw(SubarchitectureComponentException(__HERE__,"Task registration connection to goal manager has not been set up."));
//     }

//   }

//   void 
//   ManagedComponent::GMResponseHandler::nextInQueue(cdl::TaskManagementResult & _data) {
    
//     if(_data.m_decision != cdl::GOAL_WAITING) {
      
//       std::string taskID(_data.m_id);
//       m_pGDP->lockComponent();
	  
//       try {

// 	if(_data.m_decision == cdl::GOAL_ADOPTED) {
// 	  //m_pGDP->logTaskAdopted(taskID);
// 	  m_pGDP->taskAdopted(taskID);	    
// 	}
// 	else if(_data.m_decision == cdl::GOAL_REJECTED) {
// 	  m_pGDP->taskRejected(taskID);	    
// 	}	  
//       }	  
//       catch(const DoesNotExistOnWMException &e) {
// 	m_pGDP->println("ManagedComponent::taskAdopted DoesNotExistOnWMException caught");
// 	m_pGDP->println("The task involved in this error was: " + taskID);	  		
// 	m_pGDP->println("The address involved in this error was: " + 
// 			string(e.address().m_id) + " in " + 
// 			string(e.address().m_subarchitecture));	  		
// 	m_pGDP->println(e.what());
// 	std::abort();
//       }
//       catch(const CASTException &e) {
// 	m_pGDP->println("ManagedComponent::taskAdopted DoesNotExistOnWMException caught");
// 	m_pGDP->println("The task involved in this error was: " + taskID);	  		
// 	m_pGDP->println(e.what());
// 	std::abort();
//       }
//       catch(const BALTException &e) {
// 	m_pGDP->println("ManagedComponent::taskAdopted BALTException caught");
// 	m_pGDP->println("The task involved in this error was: " + taskID);	  		
// 	m_pGDP->println(e.what());
// 	std::abort();
//       }
//       catch(const CORBA::BAD_PARAM &e) {
// 	m_pGDP->println("ManagedComponent::taskAdopted CORBA::BAD_PARAM caught");		
// 	m_pGDP->println("The task involved in this error was: " + taskID);	  		
// 	m_pGDP->println(e._name());
// 	std::abort();
//       }
//       catch(const CORBA::Exception &e) {
// 	m_pGDP->println("ManagedComponent::taskAdopted CORBA::Exception caught");
// 	m_pGDP->println("The task involved in this error was: " + taskID);	  		
// 	m_pGDP->println(e._name());
// 	std::abort();
//       }
//       catch(const std::exception &e) {
// 	m_pGDP->println("ManagedComponent::taskAdopted std::exception caught");
// 	m_pGDP->println("The task involved in this error was: " + taskID);	  		
// 	m_pGDP->println(e.what());
// 	std::abort();
//       }
//       catch(...) {
// 	m_pGDP->println("ManagedComponent::taskAdopted unknown exception caught");
// 	m_pGDP->println("The task involved in this error was: " + taskID);	  		
// 	std::abort();
//       }
	    
	    
//       m_pGDP->unlockComponent();
	 
//       //wake waiting threads
//       m_pGDP->m_taskNotificationMutex.lock();
//       //increment notification count
//       m_pGDP->m_notifications++;    
//       m_pGDP->m_pTaskNotificationCondition->broadcast();
//       m_pGDP->m_taskNotificationMutex.unlock();	    	    
//     }
//   }



} //namespace cast 
